app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/ThreadController.kt

```diff
@@ -118,32 +100,53 @@ object ThreadController {
 
     //region Mark as seen/unseen
     fun toggleSeenStatus(thread: Thread) {
-        val folderId = MainViewModel.currentFolderId.value!!
+        if (thread.unseenMessagesCount == 0) markAsUnseen(thread) else markAsSeen(thread)
+    }
+
+    private fun markAsUnseen(thread: Thread) {
+
+        val mailboxUuid = MailboxController.getCurrentMailboxUuid() ?: return
+
         RealmDatabase.mailboxContent().writeBlocking {
-            if (thread.unseenMessagesCount == 0) {
-                markAsUnseen(thread, folderId, realm = this)
-            } else {
-                markAsSeen(thread, folderId, realm = this)
+            val latestThread = findLatest(thread) ?: return@writeBlocking
+            val uid = getThreadLastMessageUid(latestThread)
+            with(ApiRepository.markMessagesAsUnseen(mailboxUuid, uid)) {
+                if (isSuccess()) markThreadAsUnseen(latestThread)
             }
         }
     }
 
-    private fun markAsUnseen(thread: Thread, folderId: String, realm: MutableRealm) {
-        val latestThread = realm.findLatest(thread) ?: return
-        val uid = getThreadLastMessageUid(latestThread)
-        with(ApiRepository.markMessagesAsUnseen(latestThread.mailboxUuid, uid)) {
-            if (isSuccess()) realm.markThreadAsUnseen(latestThread, folderId)
+    fun markAsSeen(thread: Thread) {
+        if (thread.unseenMessagesCount == 0) return
+
+        val mailboxUuid = MailboxController.getCurrentMailboxUuid() ?: return
+
+        RealmDatabase.mailboxContent().writeBlocking {
+            val latestThread = findLatest(thread) ?: return@writeBlocking
+            val uids = getThreadUnseenMessagesUids(latestThread)
+            with(ApiRepository.markMessagesAsSeen(mailboxUuid, uids)) {
+                if (isSuccess()) markThreadAsSeen(latestThread)
+            }
         }
     }
 
-    fun markAsSeen(thread: Thread, folderId: String, realm: MutableRealm) {
-        if (thread.unseenMessagesCount == 0) return
+    private fun MutableRealm.markThreadAsUnseen(thread: Thread) {
+        val message = thread.messages.getLastMessageToExecuteAction()
+        message.seen = false
+        thread.unseenMessagesCount++
+
+        incrementFolderUnreadCount(message.folderId, 1)
+    }
 
-        val latestThread = realm.findLatest(thread) ?: return
-        val uids = getThreadUnseenMessagesUids(latestThread)
-        with(ApiRepository.markMessagesAsSeen(latestThread.mailboxUuid, uids)) {
-            if (isSuccess()) realm.markThreadAsSeen(latestThread, folderId)
+    private fun MutableRealm.markThreadAsSeen(thread: Thread) {
+        thread.messages.forEach {
+            if (!it.seen) {
+                incrementFolderUnreadCount(it.folderId, -1)
+                it.seen = true
+            }
         }
```

Comment:
You could call `incrementFolderUnreadCount` after the loop

Reply:
The `folderId` is different for each `Message`, so we need to do it in the loop.